### ISSUE TITLE
Performance#HashEachMethods relies on hardcoded method name

### ISSUE THREAD

**Issue by aquach** [bbatsov/rubocop#4368]
It appears that the Performance#HashEachMethods cop has a `(send _ :hash)` matcher, which means that it'll only work on method invocations of a method called `hash` as opposed to all hashes.

Since Rubocop can't infer the type, it seems like it should look for all invocations of `keys.each` or `values.each` and suggest `each_key` or `each_value`.

As for detecting `hash.each { |k, _v| }`, I'm not sure if that's possible to reliably detect with Rubocop, but I'm definitely not an expert here.

--------

## Expected behavior

Catches these issues:

```
h = {}
h.keys.each { |k| p k }
h.values.each { |v| p v }
h.each { |k, _v| p k }
h.each { |_k, v| p v }

hash = {}
hash.keys.each { |k| p k }
hash.values.each { |v| p v }
hash.each { |k, _v| p k }
hash.each { |_k, v| p v }

def b
  {}
end

b.keys.each { |k| p k }
b.values.each { |v| p v }
b.each { |k, _v| p k }
b.each { |_k, v| p v }
```

## Actual behavior

None of the above are caught.

## Steps to reproduce the problem

Paste the expected behavior and run Rubocop against it on master. There should be no `Performance#HashEachMethods` offenses.

## RuboCop version

```
0.48.1 (using Parser 2.4.0.0, running on ruby 2.2.2 x86_64-linux)
```

**Comment by Drenmi** [comment]
> It appears that the Performance#HashEachMethods cop has a `(send _ :hash)` matcher, which means that it'll only work on method invocations of a method called hash as opposed to all hashes.

You are right. Having the cop match a method send `hash` seems very arbitrary. It depends on some very vague assumption about the code, and arguably encourages bad naming. 🙂 

> Since Rubocop can't infer the type, it seems like it should look for all invocations of `keys.each` or `values.each` and suggest `each_key` or `each_value`.

I agree. In the unlikely event that someone defines their own object which responds to `#keys` and `#values` with enumerators, they should be encouraged to also implement `#each_key` and `#each_value` for the sake of polymorphism.

> As for detecting `hash.each { |k, _v| }`, I'm not sure if that's possible to reliably detect with Rubocop, but I'm definitely not an expert here.

This could be a reasonable extension to the cop. Would you be interested in making a PR? 🙂

**Comment by smakagon** [comment]
It's a really good catch. Cop works only if variable named `hash`.

@Drenmi I like your idea:
> I agree. In the unlikely event that someone defines their own object which responds to #keys and #values with enumerators, they should be encouraged to also implement #each_key and #each_value for the sake of polymorphism.

In that case we would need to catch code like this `*.keys.each` and `*.values.each`, and suggest to change it to `*.each_key` and `*.each_value`. 
How many "false positives" we would get in that case? Is that safe enough to introduce such general rule? @bbatsov what do you think about that?

**Comment by bbatsov** [comment]
> How many "false positives" we would get in that case? Is that safe enough to introduce such general rule? @bbatsov what do you think about that?

Probably not that much. I'm fine with this approach.

**Comment by gohdaniel15** [comment]
I think I'll give this a go
